File: clients/php/Model/Quest.php
<?php

namespace Model {
    require_once 'Model/Vec2Int.php';
    require_once 'Stream.php';

    /**
     * A delivery quest
     */
    class Quest
    {
        /**
         * Cell where to pick delivery from
         */
        public \Model\Vec2Int $pickupCell;
        /**
         * Cell to drop the delivery at
         */
        public \Model\Vec2Int $dropCell;
        /**
         * Score for completing the quest
         */
        public int $score;
    
        function __construct(\Model\Vec2Int $pickupCell, \Model\Vec2Int $dropCell, int $score)
        {
            $this->pickupCell = $pickupCell;
            $this->dropCell = $dropCell;
            $this->score = $score;
        }
    
        /**
         * Read Quest from input stream
         */
        public static function readFrom(\InputStream $stream): Quest
        {
            $pickupCell = \Model\Vec2Int::readFrom($stream);
            $dropCell = \Model\Vec2Int::readFrom($stream);
            $score = $stream->readInt64();
            return new Quest($pickupCell, $dropCell, $score);
        }
        
        /**
         * Write Quest to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $this->pickupCell->writeTo($stream);
            $this->dropCell->writeTo($stream);
            $stream->writeInt64($this->score);
        }
    }
}

File: clients/rust/src/model/quest.rs
use super::*;

/// A delivery quest
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Quest {
    /// Cell where to pick delivery from
    pub pickup_cell: model::Vec2I32,
    /// Cell to drop the delivery at
    pub drop_cell: model::Vec2I32,
    /// Score for completing the quest
    pub score: i64,
}

impl trans::Trans for Quest {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.pickup_cell.write_to(writer)?;
        self.drop_cell.write_to(writer)?;
        self.score.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let pickup_cell: model::Vec2I32 = trans::Trans::read_from(reader)?;
        let drop_cell: model::Vec2I32 = trans::Trans::read_from(reader)?;
        let score: i64 = trans::Trans::read_from(reader)?;
        Ok(Self {
            pickup_cell,
            drop_cell,
            score,
        })
    }
}